### ISSUE TITLE
check_http reporting a negative time

### ISSUE THREAD

**Issue by halindrome** [monitoring-plugins/monitoring-plugins#1792]
I know this should be impossible.  I am looking at the source code, and I cannot imagine how this could happen.  However, many times per day on various machines we monitor we get a negative response time reported - like this:

`HTTP CRITICAL: HTTP/1.1 200 OK - 275 bytes in -0.317 second response time`

This is running check_http on the same machine that the HTTP server is on, using arguments like this:

`check_http -H localhost -w 5 -c 10`

The machine in question is a raspberry pi running a fairly recent release of raspbian.  check_http reports that it is version 2.2. The HTTP server that is being hit is actually running in a docker container - but based upon the source code I cannot imagine how this would matter.

We are using NCSA and passive checks, so my workaround for this is to add some code to ignore negative values in the wrapper script, but that seems silly.  My question is how is it possible for check_http to call the deltime function in utils.c and get a negative value?  Or, if it is not getting a negative value from deltime, how is it possible that dividing the returned value by 1000000 results in a negative value?  Could there be some issue with using a long when the underlying system uses some other type in the time structures?  Although in the case of the Pi, it would seem that the time structures use longs...

Anyway, I am at a loss!

**Comment by RincewindsHat** [comment]
Hi @halindrome ,
This indeed seems pretty weird. I took a quick look at the code and came to the same conclusion as you did. For the sake of completeness, could you add the version of the Monitoring Plugins or `check_http` to be specific (`check_http -V`) and the architecture of your system (possibly in `uname -a` or something like that)

**Comment by halindrome** [comment]
uname -a output:

`Linux nci-elz 4.19.97-v7l+ #1294 SMP Thu Jan 30 13:21:14 GMT 2020 armv7l GNU/Linux`

This is a relatively recent Raspberry Pi.  All of the devices in question are the same, although with slight different kernels.  THe HTTP server being hit (which should not matter) is in a docker container on the same machine and listening on the host's port 80.

**Comment by RincewindsHat** [comment]
In the mean time, is `check_curl` available on those machines and could you test, whether the same problem occurs with that?
It should work as a drop in replacement in most of the cases.

**Comment by RincewindsHat** [comment]
Another inquiry: Is it possible, that the OS corrects the clock regularly? I think the implementation here does not use a strictly monotonic clock, so, if NTP or any other service updates the clock, this might happen.
Could you maybe take a look at the system and try to find out, how much often the time is changed?

**Comment by halindrome** [comment]
Yes, the clock is updated hourly in most cases.  These updates do not occur at the same time the checks are being run, though.  Are you suggesting it is possible the clock in the Docker container gets out of sync with the clock of the host?

**Comment by RincewindsHat** [comment]
Hard to say, I have no idea to be honest. But, judging from the code, the clock used here is not strictly monotonic, so this might be a reason.
The other possibility might be weird number converting in C.
I am mostly guessing here.
